proc/threads.c

```diff
@@ -2086,120 +2088,197 @@ void proc_threadsDump(u8 priority)
 }
 
 
-int proc_threadsList(int n, threadinfo_t *info)
+static inline void _proc_makeProcessName(thread_t *thread, char *name, int sz)
 {
-	int i = 0, argc;
-	unsigned int len, space;
-	thread_t *t;
+	int i, len, left = sz;
+
+	if (thread == NULL || name == NULL || sz == 0) {
+		return;
+	}
+
+	hal_memset(name, 0, sz);
+	if (thread->process != NULL && thread->process->argv != NULL) {
+		for (i = 0; thread->process->argv[i] != NULL; i++) {
+			if (left <= 0) {
+				break;
+			}
+			len = min(hal_strlen(thread->process->argv[i]) + 1U, left);
+			hal_memcpy(name, thread->process->argv[i], len);
+			name[len - 1U] = ' ';
+			name += len;
+			left -= len;
+		}
+		*(name - 1) = '\0';
+	}
+	else if (thread->process != NULL && thread->process->path != NULL) {
+		len = min(sz, hal_strlen(thread->process->path) + 1U);
+		hal_memcpy(name, thread->process->path, len);
+		name[len - 1] = '\0';
+	}
+	else if (thread->process == NULL) {
+		hal_memcpy(name, "[idle]", sizeof("[idle]"));
+	}
+}
+
+
+static inline int _proc_calculateVmem(thread_t *thread)
+{
+	int vmem = 0;
 	map_entry_t *entry;
+
+#ifdef NOMMU
+	if (thread->process != NULL) {
+		entry = thread->process->entries;
+		if (entry != NULL) {
+			do {
+				vmem += (int)entry->size;
+				entry = entry->next;
+			} while (entry != thread->process->entries);
+		}
+	}
+#else /* !NOMMU */
 	vm_map_t *map;
-	time_t now;
-	char *name;
-	spinlock_ctx_t sc;
+	spinlock_ctx_t ctx;
 
-	(void)proc_lockSet(&threads_common.lock);
+	hal_spinlockSet(&threads_common.spinlock, &ctx);
 
-	t = lib_treeof(thread_t, idlinkage, lib_rbMinimum(threads_common.id.root));
+	if (thread->process != NULL) {
+		map = thread->process->mapp;
+	}
+	else {
+		map = threads_common.kmap;
+	}
 
-	while (i < n && t != NULL) {
-		if (t->process != NULL) {
-			info[i].pid = process_getPid(t->process);
-			/* TODO: info[i].ppid = t->process->parent != NULL ? t->process->parent->id : 0; */
-			info[i].ppid = 0;
-		}
-		else {
-			info[i].pid = 0;
-			info[i].ppid = 0;
+	hal_spinlockClear(&threads_common.spinlock, &ctx);
+
+	if (map != NULL) {
+		(void)proc_lockSet(&map->lock);
+
+		entry = lib_treeof(map_entry_t, linkage, lib_rbMinimum(map->tree.root));
+		while (entry != NULL) {
+			vmem += (int)entry->size;
+			entry = lib_treeof(map_entry_t, linkage, lib_rbNext(&entry->linkage));
 		}
 
-		hal_spinlockSet(&threads_common.spinlock, &sc);
-		info[i].tid = (unsigned int)proc_getTid(t);
-		info[i].priority = (int)t->priorityBase;
-		info[i].state = (int)t->state;
+		(void)proc_lockClear(&map->lock);
+	}
+#endif
 
-		now = _proc_gettimeRaw();
-		if (now != t->startTime) {
-			info[i].load = (int)((t->cpuTime * 1000) / (now - t->startTime));
+	return vmem;
+}
+
+
+/* call with threads_common.lock set */
+void proc_threadInfo(thread_t *thread, unsigned int flags, threadinfo_t *info)
```

Comment:
If expected to be called with a lock, name it `_proc_threadInfo` (`_` implies need for external synchronization)